cmd/queuebench/main.go

```diff
@@ -0,0 +1,209 @@
+
+	apmqueue "github.com/elastic/apm-queue"
+	"github.com/elastic/apm-queue/kafka"
+)
+
+const namespace = "queuebench"
+
+func main() {
+	// NOTE: intercept any panic and terminate gracefully
+	// This allows using log.Panic in main which triggers
+	// deferred functions (whereas log.Fatal don't).
+	defer func() {
+		if r := recover(); r != nil {
+			log.Fatal(r)
+		}
+	}()
+
+	cfg := config{}
+	cfg.Parse()
+	log.Printf("parsed config: %+v\n", cfg)
+
+	log.Println("prep logger")
+	var logger *zap.Logger
+	var err error
+	if cfg.verbose {
+		logger, err = zap.NewDevelopment()
+		if err != nil {
+			log.Fatalf("cannot create zap logger: %s", err)
+		}
+	} else {
+		logger = zap.NewNop()
+	}
+
+	log.Println("prep MeterProvider")
+	rdr := sdkmetric.NewManualReader()
+	mp := sdkmetric.NewMeterProvider(
+		sdkmetric.WithReader(rdr),
+	)
+
+	ctx := context.Background()
+	run := time.Now().Unix()
+	log.Printf("running bench run: %d", run)
+
+	bench := bench{
+		Brokers:         []string{cfg.broker},
+		ConsumerGroupID: fmt.Sprintf("queuebench-%d", run),
+		Logger:          logger,
+		Partitions:      cfg.partitions,
+		TopicNamespace:  namespace,
+		Topics: []apmqueue.Topic{
+			apmqueue.Topic(fmt.Sprintf("run-%d", run)),
+		},
+
+		mp: mp,
+		tp: trace.NewNoopTracerProvider(),
+	}
+
+	log.Println("running benchmark setup")
+	if err = bench.Setup(ctx); err != nil {
+		log.Panicf("benchmark setup failed: %s", err)
+	}
+	teardown := func() {
+		log.Println("running benchmark teardown")
+		if err := bench.Teardown(ctx); err != nil {
+			log.Panicf("benchmark teardown failed: %s", err)
+		}
+
+	}
+	defer teardown()
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		for range c {
+			teardown()
+		}
+	}()
+
+	start := time.Now()
+	log.Println("==> running benchmark")
+
+	log.Println("start consuming")
+	var consumptionduration time.Duration
+	go func() {
+		consumptionstart := time.Now()
+		if err := bench.c.Run(ctx); err != nil {
+			log.Panicf("consumer run ended with an error: %s", err)
+		}
+		consumptionduration = time.Since(consumptionstart)
+	}()
+
+	log.Printf("start producing, will produce for %s", cfg.duration)
+	productionstart := time.Now()
+	if err := produce(ctx, bench.p, bench.Topics[0], cfg.eventSize, cfg.duration); err != nil {
+		log.Panicf("error while producing records: %s", err)
+	}
+	productionduration := time.Since(productionstart)
+
+	log.Println("production ended")
+
+	log.Println("waiting for consumer to fetch all records")
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+	timer := time.NewTimer(cfg.timeout)
+	defer timer.Stop()
+
+	var rm metricdata.ResourceMetrics
+	totalproduced := int64(0)
+	totalconsumed := int64(0)
+wait:
+	for {
+		ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+		defer cancel()
+		select {
+		case <-timer.C:
+			log.Println("reached timeout, moving on")
+			break wait
+		case <-ticker.C:
+			rdr.Collect(ctx, &rm)
```

Comment:
Should we be checking the result here? Otherwise how would we react to the context being cancelled?

Reply:
otel code documentation says:
> 	  // Collect gathers and returns all metric data related to the Reader from
>	  // the SDK and stores it in out. An error is returned if this is called
>	  // after Shutdown or if out is nil.
>	  Collect(ctx context.Context, rm *metricdata.ResourceMetrics) error

I don't think is possible to call it after `Shutdown` in this case, but I added error handling for the `nil` case.

I added a select case to intercept `ctx.Done` to detect context cancellation.